File: clients/typescript/src/transport.ts
import { NovamemError } from "./errors.js";

/** Bounds a single call when no timeout is given. Sized for the slowest
 * operation: capture embeds the content server-side before it answers. */
export const DEFAULT_TIMEOUT_MS = 15_000;
/** Responses larger than this are rejected rather than buffered. */
export const MAX_RESPONSE_BYTES = 8 << 20;

export interface ClientOptions {
  /** The service root, e.g. "https://novamem.example.com". */
  baseUrl: string;
  /** The user's `nm_…` bearer. Never included in an error. */
  token: string;
  /** Bounds each call, even one given no AbortSignal. Default 15 000. */
  timeoutMs?: number;
  /** Injected fetch, for tests and custom agents. Default: global fetch. */
  fetch?: typeof fetch;
}

export interface CallOptions {
  /** Aborting it ends the call with a `canceled` NovamemError. */
  signal?: AbortSignal;
}

interface Request {
  body?: unknown;
  query?: Record<string, string | number | boolean | undefined | null>;
  expectBody?: boolean;
  signal?: AbortSignal;
}

/** One function makes every request, so failures are classified once. */
export class Transport {
  readonly baseUrl: string;
  readonly timeoutMs: number;
  readonly #token: string;
  readonly #fetch: typeof fetch;

  constructor(o: ClientOptions) {
    const base = (o?.baseUrl ?? "").trim().replace(/\/+$/, "");
    let ok = false;
    try {
      const u = new URL(base);
      ok = (u.protocol === "http:" || u.protocol === "https:") && u.host !== "";
    } catch {
      ok = false;
    }
    // Names the field, never the value: a token pasted into the wrong
    // option must not end up in a log line.
    if (!ok)
      throw new NovamemError({
        op: "config",
        message: "baseUrl is not an absolute http(s) URL",
      });
    if (!(o.token ?? "").trim())
      throw new NovamemError({ op: "config", message: "token is required" });
    this.baseUrl = base;
    this.#token = o.token;
    this.timeoutMs =
      o.timeoutMs && o.timeoutMs > 0 ? o.timeoutMs : DEFAULT_TIMEOUT_MS;
    this.#fetch = o.fetch ?? globalThis.fetch;
  }

  toJSON(): object {
    return {
      baseUrl: this.baseUrl,
      token: "[redacted]",
      timeoutMs: this.timeoutMs,
    };
  }

  [Symbol.for("nodejs.util.inspect.custom")](): string {
    return `Transport { baseUrl: '${this.baseUrl}', token: [redacted] }`;
  }

  #redact(s: string): string {
    return s.split(this.#token).join("[redacted]");
  }

  async call(
    op: string,
    method: string,
    path: string,
    r: Request = {}
  ): Promise<any> {
    let url = this.baseUrl + path;
    const q = new URLSearchParams();
    for (const [k, v] of Object.entries(r.query ?? {}))
      if (v !== undefined && v !== null && v !== "") q.set(k, String(v));
    if ([...q].length) url += "?" + q.toString();
    const headers: Record<string, string> = {
      Accept: "application/json",
      Authorization: `Bearer ${this.#token}`,
    };
    let body: string | undefined;
    if (r.body !== undefined) {
      body = JSON.stringify(r.body);
      headers["Content-Type"] = "application/json";
    }

    // Every call is bounded, even one given no signal. The timeout and the
    // caller's signal are told apart afterwards: a timeout means "could not
    // look", a cancel means the caller walked away.
    const timeout = new AbortController();
    const timer = setTimeout(() => timeout.abort(), this.timeoutMs);
    const both = new AbortController();
    const onAbort = () => both.abort();
    timeout.signal.addEventListener("abort", onAbort);
    r.signal?.addEventListener("abort", onAbort);
    if (r.signal?.aborted) both.abort();

    let status = 0;
    let raw: Uint8Array;
    try {
      const res = await this.#follow(url, {
        method,
        headers,
        body,
        signal: both.signal,
      });
      status = res.status;
      raw = await readCapped(res);
    } catch (err) {
      if (r.signal?.aborted)
        throw new NovamemError({ op, message: "canceled", canceled: true });
      if (timeout.signal.aborted)
        throw new NovamemError({
          op,
          message: "timed out",
          unavailable: true,
          retryable: true,
        });
      if (err instanceof TooLarge) {
        throw new NovamemError({
          op,
          statusCode: status,
          message: "response body exceeds 8 MiB",
          unavailable: true,
        });
      }
      const cause =
        (err as any)?.cause?.code ?? (err as any)?.message ?? String(err);
      throw new NovamemError({
        op,
        message: this.#redact(`unreachable: ${cause}`),
        unavailable: true,
        retryable: true,
      });
    } finally {
      clearTimeout(timer);
      r.signal?.removeEventListener("abort", onAbort);
    }

    const text = new TextDecoder().decode(raw);
    if (status < 200 || status >= 300) throw this.#httpError(op, status, text);
    if (r.expectBody === false) return undefined;
    // A 2xx with no body is not the contract: decoding it into a default
    // would tell a forget caller the delete happened.
    if (!text.trim())
      throw new NovamemError({
        op,
        statusCode: status,
        message: "empty response body",
        unavailable: true,
      });
    try {
      return JSON.parse(text);
    } catch {
      // In practice a proxy's HTML error page: we never reached a working
      // novamem. Not retryable — the same request parses the same way.
      throw new NovamemError({
        op,
        statusCode: status,
        message: "malformed response body",
        unavailable: true,
      });
    }
  }

  /** Follows up to 5 redirects itself, so the bearer never reaches another
   * origin: a cross-origin hop drops Authorization, as Go's http.Client
   * does. 303, and 301/302 after a POST, continue as a bodyless GET. */
  async #follow(
    url: string,
    init: RequestInit & { headers: Record<string, string> }
  ): Promise<Response> {
    let current = url;
    let req = init;
    for (let hops = 0; ; hops++) {
      const res = await this.#fetch(current, { ...req, redirect: "manual" });
      const location = res.headers.get("location");
      if (
        res.status < 300 ||
        res.status >= 400 ||
        res.status === 304 ||
        !location ||
        hops === 5
      )
        return res;
      await res.body?.cancel();
      const next = new URL(location, current);
      const headers = { ...req.headers };
      if (next.origin !== new URL(current).origin) delete headers.Authorization;
      const toGet =
        res.status === 303 ||
        ((res.status === 301 || res.status === 302) && req.method === "POST");
      if (toGet) delete headers["Content-Type"];
      req = {
        ...req,
        headers,
        method: toGet ? "GET" : req.method,
        body: toGet ? undefined : req.body,
      };
      current = next.toString();
    }
  }

  #httpError(op: string, status: number, text: string): NovamemError {
    let message = "";
    let code = "";
    try {
      const p = JSON.parse(text);
      if (p && typeof p === "object" && p.error) {
        message = String(p.error);
        code = p.code ? String(p.code) : "";
      }
    } catch {
      // not JSON: fall through to the raw text
    }
    if (!message) {
      const t = text.trim();
      message = t.length > 256 ? t.slice(0, 256) + "…" : t;
    }
    // The server's message is quoted verbatim; a server echoing the token
    // back would otherwise launder it into the caller's logs.
    message = this.#redact(message);
    code = this.#redact(code);
    const unavailable = status >= 500 || status === 429;
    return new NovamemError({
      op,
      statusCode: status,
      code,
      message,
      unavailable,
      retryable: unavailable,
    });
  }
}

class TooLarge extends Error {}

async function readCapped(res: Response): Promise<Uint8Array> {
  if (!res.body) return new Uint8Array();
  const reader = res.body.getReader();
  const chunks: Uint8Array[] = [];
  let size = 0;
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    size += value.byteLength;
    if (size > MAX_RESPONSE_BYTES) {
      await reader.cancel();
      throw new TooLarge();
    }
    chunks.push(value);
  }
  const out = new Uint8Array(size);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.byteLength;
  }
  return out;
}

File: clients/rust/src/transport.rs
//! One function makes every request, so the classification of failures is
//! written exactly once.

use std::fmt;
use std::time::Duration;

use reqwest::Method;
use serde_json::Value;

use crate::Error;

/// Bounds a single call when no timeout is given. Sized for the slowest
/// operation: capture embeds the content server-side before it answers.
pub const DEFAULT_TIMEOUT: Duration = Duration::from_secs(15);

/// Responses larger than this are rejected rather than buffered.
pub const MAX_RESPONSE_BYTES: usize = 8 << 20;

/// Everything a client needs, all of it injected: nothing is read from the
/// environment, so two clients — or one pointed at a test server — can
/// live in one process.
#[derive(Clone, Default)]
pub struct Config {
    /// The service root, e.g. `https://novamem.example.com`.
    pub base_url: String,
    /// The user's `nm_…` bearer. Never included in an error or in `Debug`.
    pub token: String,
    /// Bounds each call. `None` means [`DEFAULT_TIMEOUT`].
    pub timeout: Option<Duration>,
    /// An injected client, e.g. with a custom root store. Its own redirect
    /// policy then applies.
    pub http: Option<reqwest::Client>,
}

impl fmt::Debug for Config {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Config")
            .field("base_url", &self.base_url)
            .field("token", &"[redacted]")
            .field("timeout", &self.timeout)
            .finish()
    }
}

#[derive(Clone)]
pub(crate) struct Transport {
    base: String,
    token: String,
    timeout: Duration,
    http: reqwest::Client,
}

impl fmt::Debug for Transport {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Transport")
            .field("base", &self.base)
            .field("token", &"[redacted]")
            .finish()
    }
}

/// A successful response: its status, kept so a body that decodes as JSON
/// but not as the expected type can still report it.
pub(crate) struct Resp {
    pub status: u16,
    pub body: Value,
}

pub(crate) struct Call<'a> {
    pub op: &'a str,
    pub method: Method,
    pub path: String,
    pub body: Option<Value>,
    pub query: Vec<(&'a str, String)>,
    pub expect_body: bool,
}

impl<'a> Call<'a> {
    pub fn new(op: &'a str, method: Method, path: impl Into<String>) -> Self {
        Call {
            op,
            method,
            path: path.into(),
            body: None,
            query: Vec::new(),
            expect_body: true,
        }
    }

    pub fn body(mut self, body: Value) -> Self {
        self.body = Some(body);
        self
    }

    pub fn query(mut self, k: &'a str, v: Option<impl ToString>) -> Self {
        if let Some(v) = v {
            let v = v.to_string();
            if !v.is_empty() {
                self.query.push((k, v));
            }
        }
        self
    }

    pub fn no_body(mut self) -> Self {
        self.expect_body = false;
        self
    }
}

impl Transport {
    pub fn new(cfg: Config) -> Result<Self, Error> {
        let base = cfg.base_url.trim().trim_end_matches('/').to_string();
        let ok = reqwest::Url::parse(&base)
            .map(|u| {
                (u.scheme() == "http" || u.scheme() == "https")
                    && u.host_str().is_some_and(|h| !h.is_empty())
            })
            .unwrap_or(false);
        // Names the field, never the value: a token pasted into the wrong
        // field must not end up in a log line.
        if !ok {
            return Err(Error::new(
                "config",
                "base_url is not an absolute http(s) URL",
            ));
        }
        if cfg.token.trim().is_empty() {
            return Err(Error::new("config", "token is required"));
        }
        let http = match cfg.http {
            Some(h) => h,
            // reqwest drops Authorization on a redirect to another host, as
            // Go's http.Client does (tests/redirect.rs).
            None => reqwest::Client::builder()
                .use_rustls_tls()
                .build()
                .map_err(|e| Error::new("config", format!("building the HTTP client: {e}")))?,
        };
        Ok(Transport {
            base,
            token: cfg.token,
            timeout: cfg.timeout.unwrap_or(DEFAULT_TIMEOUT),
            http,
        })
    }

    fn redact(&self, s: &str) -> String {
        s.replace(&self.token, "[redacted]")
    }

    /// Performs one request and returns its status and decoded JSON body
    /// (`Null` when the call expects none). Every call is bounded by the
    /// timeout.
    pub async fn call(&self, c: Call<'_>) -> Result<Resp, Error> {
        let op = c.op;
        match tokio::time::timeout(self.timeout, self.exchange(&c)).await {
            Err(_) => Err(Error::unavailable(op, "timed out", true)),
            Ok(Err(e)) => Err(e),
            Ok(Ok((status, raw))) => Ok(Resp {
                status,
                body: self.decode(op, status, &raw, c.expect_body)?,
            }),
        }
    }

    async fn exchange(&self, c: &Call<'_>) -> Result<(u16, Vec<u8>), Error> {
        let mut req = self
            .http
            .request(c.method.clone(), format!("{}{}", self.base, c.path))
            .header(reqwest::header::ACCEPT, "application/json")
            .bearer_auth(&self.token);
        if !c.query.is_empty() {
            req = req.query(&c.query);
        }
        if let Some(b) = &c.body {
            req = req
                .header(reqwest::header::CONTENT_TYPE, "application/json")
                .body(
                    serde_json::to_vec(b)
                        .map_err(|e| Error::new(c.op, format!("encode request: {e}")))?,
                );
        }
        let mut resp = req
            .send()
            .await
            .map_err(|e| self.transport_error(c.op, &e))?;
        let status = resp.status().as_u16();
        let mut raw = Vec::new();
        while let Some(chunk) = resp
            .chunk()
            .await
            .map_err(|e| self.transport_error(c.op, &e))?
        {
            raw.extend_from_slice(&chunk);
            if raw.len() > MAX_RESPONSE_BYTES {
                return Err(
                    Error::unavailable(c.op, "response body exceeds 8 MiB", false)
                        .with_status(status),
                );
            }
        }
        Ok((status, raw))
    }

    fn transport_error(&self, op: &str, e: &reqwest::Error) -> Error {
        if e.is_timeout() {
            return Error::unavailable(op, "timed out", true);
        }
        // Refused dial, DNS, reset, TLS: the host could not be consulted.
        let mut msg = e.to_string();
        let mut src = std::error::Error::source(e);
        while let Some(s) = src {
            msg = format!("{msg}: {s}");
            src = s.source();
        }
        Error::unavailable(op, self.redact(&format!("unreachable: {msg}")), true)
    }

    fn decode(&self, op: &str, status: u16, raw: &[u8], expect_body: bool) -> Result<Value, Error> {
        if !(200..300).contains(&status) {
            return Err(self.http_error(op, status, raw));
        }
        if !expect_body {
            return Ok(Value::Null);
        }
        if raw.iter().all(u8::is_ascii_whitespace) {
            // A 2xx with no body is not the contract: decoding it into a
            // default would tell a forget caller the delete happened.
            return Err(Error::unavailable(op, "empty response body", false).with_status(status));
        }
        // In practice a proxy's HTML error page: we never reached a working
        // novamem. Not retryable — the same request parses the same way.
        serde_json::from_slice(raw).map_err(|_| {
            Error::unavailable(op, "malformed response body", false).with_status(status)
        })
    }

    fn http_error(&self, op: &str, status: u16, raw: &[u8]) -> Error {
        let (mut message, mut code) = (String::new(), String::new());
        if let Ok(Value::Object(m)) = serde_json::from_slice::<Value>(raw) {
            if let Some(e) = m
                .get("error")
                .and_then(Value::as_str)
                .filter(|s| !s.is_empty())
            {
                message = e.to_string();
                code = m
                    .get("code")
                    .and_then(Value::as_str)
                    .unwrap_or_default()
                    .to_string();
            }
        }
        if message.is_empty() {
            let text = String::from_utf8_lossy(raw).trim().to_string();
            message = match text.char_indices().nth(256) {
                Some((i, _)) => format!("{}…", &text[..i]),
                None => text,
            };
        }
        // The server's message and code are quoted verbatim; a server that
        // echoes the credential back would otherwise launder it into logs.
        let (message, code) = (self.redact(&message), self.redact(&code));
        let e = if status >= 500 || status == 429 {
            Error::unavailable(op, message, true)
        } else {
            Error::new(op, message)
        };
        e.with_status(status).with_code(code)
    }
}
